### PR TITLE
Refactoring overhaul to have consistent naming across API and docstrings

### PR DESCRIPTION
**Pull request by CuriousLearner**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -39,39 +39,101 @@ Usage
...
 
 Compatibility
 -------------
 - Django 2.1+
 - Django REST Framework 3.9+
 
+Contributing
+------------
+
+No code is bug-free and I'm sure this app will have bugs. If you find any bugs, please create an issue on Github.
```
Comment (GeekyShacklebolt) [inline README.rst]: A minor typo `GitHub`.

File: README.rst [hunk 2]
```diff
@@ -39,39 +39,101 @@ Usage
...
+
+- ``pre-commit-config`` to maintain code quality using black and other useful tools.
+- Docs for integration and usage in `usage.rst`_.
+- Tox for testing on `py{37}-django{20,21,22}`.
+- Travis CI for testing builds.
+
+Changed
+"""""""
+
+- Convert `*.md` docs to rSt Markup.
```
Comment (GeekyShacklebolt) [inline README.rst]: Here, `rSt` can be `reST` or `RST`.

File: phone_verify/backends/base.py [hunk 3]
```diff
@@ -56,29 +64,50 @@ def token_expired(cls, stored_verification):
...
 
-    def validate_token(self, security_code, phone_number, session_code):
+    def validate_security_code(self, security_code, phone_number, session_token):
+        """
+        A utitilty method to verify if the `security_code` entered is valid for
+        a given `phone_number` along with the `session_code` used.
+
+        :param security_code: Security code entered for verification
+        :param phone_number: Phone number to be verified
+
```
Comment (GeekyShacklebolt) [inline phone_verify/backends/base.py]: Missing `:param session_token:`.

File: README.rst [hunk 4]
```diff
@@ -39,39 +39,101 @@ Usage
...
+- Fix issue with installing required package dependecies via ``install_requires``.
+
+[0.1.1]
+^^^^^^^
+
+Added
+"""""
+
+-
```
Comment (GeekyShacklebolt) [inline README.rst]: Empty bullet `-`.

**Comment by storymode7** [comment]
I've left my comments inline. Few nitpicks here and there, rest looks good.

File: README.rst [hunk 5]
```diff
@@ -39,39 +39,101 @@ Usage
...
+    # In settings.py
     # Add settings for phone_verify to work
-    PHONE_VERIFICATION = {
-        'BACKEND': 'phone_verify.backends.twilio.TwilioBackend',
-        'TWILIO_SANDBOX_TOKEN':'123456',
-        'OPTIONS': {
-            'SID': 'fake',
-            'SECRET': 'fake',
-            'FROM': '+14755292729'
+    "PHONE_VERIFICATION" = {
```
Comment (storymode7) [inline README.rst]:
```suggestion
    PHONE_VERIFICATION = {
```
I don't think the quotes should be here.

File: README.rst (hunk 1 above)
Comment (CuriousLearner) [reply README.rst]:
```suggestion
No code is bug-free and I'm sure this app will have bugs. If you find any bugs, please create an issue on GitHub.
```

File: README.rst (hunk 2 above)
Comment (CuriousLearner) [inline README.rst]:
```suggestion
- Convert `*.md` docs to reST Markup.
```

File: phone_verify/backends/base.py (hunk 3 above)
Comment (CuriousLearner) [inline phone_verify/backends/base.py]:
```suggestion
:param session_token: Session token to identify the device
```

File: README.rst [hunk 6]
```diff
@@ -39,39 +39,101 @@ Usage
...
-        'APP_NAME': 'Phone Verify',
-        'SECURITY_CODE_EXPIRATION_TIME': 3600,  # In seconds only
-        'VERIFY_SECURITY_CODE_ONLY_ONCE': True,  # If False, then a security code can be used multiple times for verification
-    }
+        "TOKEN_LENGTH": 6,
+        "MESSAGE": "Welcome to {app}! Please use security code {security_code} to proceed.",
+        "APP_NAME": "Phone Verify",
+        "SECURITY_CODE_EXPIRATION_TIME": 3600,  # In seconds only
+        "VERIFY_SECURITY_CODE_ONLY_ONCE": False,  # If False, then a security code can be used multiple times for verification
+    },
```
Comment (storymode7) [inline README.rst]:
```suggestion
    }
```

File: phone_verify/backends/base.py [hunk 7]
```diff
@@ -11,42 +13,48 @@
...
+    @abstractmethod
     def send_bulk_sms(self, numbers, message):
         raise NotImplementedError()
 
     @classmethod
-    def generate_token(cls):
+    def generate_security_code(cls):
         """
-        Returns an unique random token
+        Returns an unique random `security_code` for given `TOKEN_LENGTH` in the settings.
```
Comment (storymode7) [inline phone_verify/backends/base.py]:
```suggestion
        Returns a unique random `security_code` for given `TOKEN_LENGTH` in the settings.
```

File: phone_verify/backends/base.py [hunk 8]
```diff
@@ -11,42 +13,48 @@
...
             "TOKEN_LENGTH", DEFAULT_TOKEN_LENGTH
         )
         return get_random_string(token_length, allowed_chars="0123456789")
 
     @classmethod
-    def generate_session_token(cls, token, phone_number):
+    def generate_session_token(cls, security_code, phone_number):
         """
-        Returns an unique random token for a particular device
+        Returns an unique random JWT token using Django's `SECRET_KEY`
```
Comment (storymode7) [inline phone_verify/backends/base.py]:
```suggestion
        Returns a unique random JWT token using Django's `SECRET_KEY`
```

File: phone_verify/backends/base.py [hunk 9]
```diff
@@ -11,42 +13,48 @@
...
+        for identifying a particular device in subsequent calls.
         """
-        data = {"device_%s_session_code" % phone_number: token}
+        data = {"device_%s_session_code" % phone_number: security_code}
```
Comment (storymode7) [inline phone_verify/backends/base.py]:
```suggestion
        data = {"device_%s_session_token" % phone_number: security_code}
```

File: phone_verify/backends/base.py [hunk 10]
```diff
@@ -56,29 +64,51 @@ def token_expired(cls, stored_verification):
...
+            session_token=session_token,
         )
-        return security_code, session_code
+        return security_code, session_token
 
-    def validate_token(self, security_code, phone_number, session_code):
+    def validate_security_code(self, security_code, phone_number, session_token):
+        """
+        A utitilty method to verify if the `security_code` entered is valid for
+        a given `phone_number` along with the `session_code` used.
```
Comment (storymode7) [inline phone_verify/backends/base.py]:
```suggestion
        a given `phone_number` along with the `session_token` used.
```

File: phone_verify/backends/twilio.py [hunk 11]
```diff
@@ -50,22 +50,11 @@ def send_bulk_sms(self, numbers, message):
         for number in numbers:
             self.send_sms(to=number, body=message, from_=self._from)
 
-    def generate_token(self):
+    def generate_security_code(self):
         """
         Returns an fixed token
```
Comment (gutsytechster) [inline phone_verify/backends/twilio.py]: I think it should be `Returns security code`

File: phone_verify/backends/twilio.py [hunk 12]
```diff
@@ -50,22 +50,11 @@ def send_bulk_sms(self, numbers, message):
...
-
-        :param number: Number of recipient
-
-        :return token: string of SHA token
-        """
-        return self.generate_token()
-
-    def validate_token(self, security_code, phone_number):
+    def validate_security_code(self, security_code, phone_number, session_token):
         return self.VALID
```
Comment (gutsytechster) [inline phone_verify/backends/twilio.py]:
```suggestion
        return self.SECURITY_CODE_VALID
```

File: phone_verify/backends/twilio.py [hunk 13]
```diff
@@ -50,22 +50,11 @@ def send_bulk_sms(self, numbers, message):
         for number in numbers:
             self.send_sms(to=number, body=message, from_=self._from)
 
-    def generate_token(self):
+    def generate_security_code(self):
         """
-        Returns an fixed token
+        Returns an fixed security code
```
Comment (gutsytechster) [inline phone_verify/backends/twilio.py]:
```suggestion
        Returns a fixed security code
```

File: README.rst (hunk 4 above)
Comment (CuriousLearner) [reply README.rst]: Will be filled later

File: phone_verify/services.py [hunk 14]
```diff
@@ -50,6 +48,25 @@ def _generate_message(self, security_code):
             security_code=security_code,
         )
 
+    def _check_required_settings(self):
+        required_settings = {
+            "BACKEND",
+            "TWILIO_SANDBOX_TOKEN",
```
Comment (CuriousLearner) [inline phone_verify/services.py]: We should remove this setting from the root of the dict

File: README.rst [hunk 15]
```diff
@@ -4,21 +4,38 @@ django-phone-verify
...
+.. image:: https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square
+   :target: http://makeapullrequest.com
+
+
+A Django app to support phone number verification using the security code sent via SMS.
 
 Salient Features
 ----------------
 
 - Let's you verify phone numbers via SMS.
```
Comment (gutsytechster) [inline README.rst]:
```suggestion
- Lets you verify phone numbers via SMS.
```
Not sure though!

File: README.rst [hunk 16]
```diff
@@ -39,39 +56,104 @@ Usage
...
+- Response for ``/api/phone/register`` contains key ``session_token`` instead of ``session_code``.
+- Request payload for ``/api/phone/verify`` now expects ``session_token`` key instead of ``session_code``.
+- Response for ``/api/phone/verify`` now sends additional response of ``Security code is already verified`` in case ``VERIFY_SECURITY_CODE_ONLY_ONCE`` is set to ``True``.
+- Rename ``otp`` to ``security_code`` in code and docs to be more consistent.
+- Rename ``BaseBackend`` status from ``VALID``, ``INVALID``, ``EXPIRED`` to ``SECURITY_CODE_VALID``, ``SECURITY_CODE_INVALID``, and ``SECURITY_CODE_EXPIRED`` respectively.
+- Rename ``session_code`` to ``session_token`` to be consistent in code and naming across the app.
+- Rename service ``send_otp_and_generate_session_code`` to ``send_security_code_and_generate_session_token``.
+- Rename method ``BaseBackend.generate_token`` to ``BaseBackend.generate_security_code``.
+- Rename method ``create_otp_and_session_token`` to ``create_security_code_and_session_token``.
+- Rename method ``BaseBackend.validate_token`` to ``validate_security_code`` with an additional parameter of ``session_token``.
```
Comment (gutsytechster) [inline README.rst]:
```suggestion
- Rename method ``BaseBackend.validate_token`` to ``BaseBackend.validate_security_code`` with an additional parameter of ``session_token``.
```

File: README.rst [hunk 17]
```diff
@@ -39,39 +56,104 @@ Usage
...
+- Fix issue with installing required package dependencies via ``install_requires``.
+
+[0.1.1]
+^^^^^^^
+
+Added
+"""""
+
+- README and documentation of API endpoints.
+- `setup.cfg` to manage coverage.
```
Comment (gutsytechster) [inline README.rst]:
```suggestion
- ``setup.cfg`` to manage coverage.
```

File: README.rst [hunk 18]
```diff
@@ -39,39 +56,104 @@ Usage
...
+- `phone_verify` app including backends, requirements, tests.
```
Comment (gutsytechster) [inline README.rst]:
```suggestion
- ``phone_verify`` app including backends, requirements, tests.
```

File: phone_verify/backends/base.py [hunk 19]
```diff
@@ -11,42 +13,48 @@
...
+    @abstractmethod
     def send_bulk_sms(self, numbers, message):
         raise NotImplementedError()
 
     @classmethod
-    def generate_token(cls):
+    def generate_security_code(cls):
         """
-        Returns an unique random token
+        Returns a unique random `security_code` for given `TOKEN_LENGTH` in the settings.
```
Comment (gutsytechster) [inline phone_verify/backends/base.py]:
```suggestion
        Returns a unique random ``security_code`` for given ``TOKEN_LENGTH`` in the settings.
```

File: phone_verify/backends/base.py [hunk 20]
```diff
@@ -11,42 +13,48 @@
...
             "TOKEN_LENGTH", DEFAULT_TOKEN_LENGTH
         )
         return get_random_string(token_length, allowed_chars="0123456789")
 
     @classmethod
-    def generate_session_token(cls, token, phone_number):
+    def generate_session_token(cls, security_code, phone_number):
         """
-        Returns an unique random token for a particular device
+        Returns a unique random JWT token using Django's `SECRET_KEY`
```
Comment (gutsytechster) [inline phone_verify/backends/base.py]:
```suggestion
        Returns a unique random JWT token using Django's ``SECRET_KEY``
```

File: phone_verify/backends/base.py [hunk 21]
```diff
@@ -11,42 +13,48 @@
...
         """
-        data = {"device_%s_session_code" % phone_number: token}
+        data = {"device_%s_session_token" % phone_number: security_code}
         return jwt.encode(data, django_settings.SECRET_KEY).decode()
 
     @classmethod
-    def token_expired(cls, stored_verification):
+    def check_security_code_expiry(cls, stored_verification):
+        """
+        Returns True if the `security_code` for the `stored_verification` is expired.
```
Comment (gutsytechster) [inline phone_verify/backends/base.py]:
```suggestion
        Returns True if the ``security_code`` for the ``stored_verification`` is expired.
```

File: phone_verify/backends/base.py [hunk 22]
```diff
@@ -56,29 +64,51 @@ def token_expired(cls, stored_verification):
 
     def create_security_code_and_session_token(self, number):
         """
-        Creates a temporary security_code inside the cache, this holds the phone number
-        as value, so that we can later check if everything is correct.
+        Creates a temporary `security_code` and `session_token` inside the DB.
```
Comment (gutsytechster) [inline phone_verify/backends/base.py]:
```suggestion
        Creates a temporary ``security_code`` and ``session_token`` inside the DB.
```

File: phone_verify/backends/base.py [hunk 23]
```diff
@@ -56,29 +64,51 @@ def token_expired(cls, stored_verification):
...
 
-        It also generates a session_code for the client to send in subsequent requests.
+        `security_code` is the code that user would enter to verify their phone_number.
```
Comment (gutsytechster) [inline phone_verify/backends/base.py]:
```suggestion
        ``security_code`` is the code that user would enter to verify their phone_number.
```

File: phone_verify/backends/base.py [hunk 24]
```diff
@@ -56,29 +64,51 @@ def token_expired(cls, stored_verification):
...
+        `session_token` is used to verify if the subsequent call for verification is
```
Comment (gutsytechster) [inline phone_verify/backends/base.py]:
```suggestion
        ``session_token`` is used to verify if the subsequent call for verification is
```

File: tests/test_api.py [hunk 25]
```diff
@@ -80,14 +80,14 @@ def test_phone_verification_with_incorrect_payload(client):
     f.create_verification(
         security_code=SECURITY_CODE,
         phone_number=PHONE_NUMBER,
-        session_code=SESSION_CODE,
+        session_token=SESSION_TOKEN,
     )
     url = reverse("phone-verify")
     # Payload with wrong session code
```
Comment (storymode7) [inline tests/test_api.py]:
```suggestion
    # Payload with wrong session token
```

File: tests/test_api.py [hunk 26]
```diff
@@ -80,14 +80,14 @@ def test_phone_verification_with_incorrect_payload(client):
...
     data = {
         "phone_number": PHONE_NUMBER,
         "security_code": SECURITY_CODE,
-        "session_code": "wrong-session-code",
+        "session_token": "wrong-session-code",
```
Comment (storymode7) [inline tests/test_api.py]:
```suggestion
        "session_token": "wrong-session-token",
```

File: phone_verify/backends/base.py [hunk 27]
```diff
@@ -56,29 +64,51 @@ def token_expired(cls, stored_verification):
...
+            session_token=session_token,
         )
-        return security_code, session_code
+        return security_code, session_token
 
-    def validate_token(self, security_code, phone_number, session_code):
+    def validate_security_code(self, security_code, phone_number, session_token):
+        """
+        A utitilty method to verify if the `security_code` entered is valid for
+        a given `phone_number` along with the `session_token` used.
```
Comment (gutsytechster) [inline phone_verify/backends/base.py]: Need to quote keywords in double backticks here as well :|

File: phone_verify/serializers.py [hunk 28]
```diff
@@ -20,24 +20,26 @@ class PhoneSerializer(serializers.Serializer):
...
+            security_code=security_code,
+            phone_number=phone_number,
+            session_token=session_token,
         )
 
         if verification is None:
             raise serializers.ValidationError(_("Security code is not valid"))
-        elif token_validatation == backend.SESSION_CODE_INVALID:
+        elif token_validatation == backend.SESSION_TOKEN_INVALID:
             raise serializers.ValidationError(_("Session Code mis-match"))
```
Comment (gutsytechster) [inline phone_verify/serializers.py]:
```suggestion
            raise serializers.ValidationError(_("Session Token mis-match"))
```

File: tests/test_api.py [hunk 29]
```diff
@@ -16,7 +16,7 @@
 
 SECURITY_CODE = "123456"
 PHONE_NUMBER = "+13478379634"
-SESSION_CODE = "phone-auth-session-code"
+SESSION_TOKEN = "phone-auth-session-code"
```
Comment (gutsytechster) [inline tests/test_api.py]:
```suggestion
SESSION_TOKEN = "phone-auth-session-token"
```

**Comment by gutsytechster** [comment]
Just a few one-word changes. Also, `usage.rst` doesn't have `VERIFY_SECURITY_CODE_ONLY_ONCE` within `PHONE_VERIFICATION` setting.

Else LGTM! :taco:

File: phone_verify/backends/base.py [hunk 30]
```diff
@@ -56,29 +64,51 @@ def token_expired(cls, stored_verification):
...
+            security_code=security_code,
+            session_token=session_token,
         )
-        return security_code, session_code
+        return security_code, session_token
 
-    def validate_token(self, security_code, phone_number, session_code):
+    def validate_security_code(self, security_code, phone_number, session_token):
+        """
+        A utitilty method to verify if the `security_code` entered is valid for
```
Comment (CuriousLearner) [inline phone_verify/backends/base.py]:
```suggestion
        A utility method to verify if the ``security_code`` entered is valid for
```

File: phone_verify/backends/base.py (hunk 27 above)
Comment (CuriousLearner) [inline phone_verify/backends/base.py]:
```suggestion
        a given ``phone_number`` along with the ``session_token`` used.
```

File: README.rst [hunk 31]
```diff
@@ -39,39 +56,104 @@ Usage
...
+        "APP_NAME": "Phone Verify",
+        "SECURITY_CODE_EXPIRATION_TIME": 3600,  # In seconds only
+        "VERIFY_SECURITY_CODE_ONLY_ONCE": False,  # If False, then a security code can be used multiple times for verification
     }
 
-- To explore more about how to use `Django Phone Verify`, have a look at `usage.rst`_
+Usage
+-----
+
+- To explore more about how to use, integrate and leverage existing functionality of `Django Phone Verify`, have a look at `usage.rst`_
```
Comment (gutsytechster) [inline README.rst]:
```suggestion
- To explore more about how to use, integrate and leverage existing functionality of ``Django Phone Verify``, have a look at `usage.rst`_
```

File: README.rst [hunk 32]
```diff
@@ -39,39 +56,104 @@ Usage
...
+
+- ``pre-commit-config`` to maintain code quality using black and other useful tools.
+- Docs for integration and usage in `usage.rst`_.
+- Tox for testing on `py{37}-django{20,21,22}`.
+- Travis CI for testing builds.
+
+Changed
+"""""""
+
+- Convert `*.md` docs to reST Markup.
```
Comment (gutsytechster) [inline README.rst]:
```suggestion
- Convert ``*.md`` docs to reST Markup.
```

File: phone_verify/backends/base.py [hunk 33]
```diff
@@ -56,29 +64,51 @@ def token_expired(cls, stored_verification):
...
+
+        :return stored_verification: Contains the verification object
+        corresponding to the phone_number if found, else None.
+        :return status: Status for the stored_verification object.
+        Can be one of the following:
+            - `BaseBackend.SECURITY_CODE_VALID`
+            - `BaseBackend.SECURITY_CODE_INVALID`
+            - `BaseBackend.SECURITY_CODE_EXPIRED`
+            - `BaseBackend.SECURITY_CODE_VERIFIED`
+            - `BaseBackend.SESSION_TOKEN_INVALID`
```
Comment (gutsytechster) [inline phone_verify/backends/base.py]: Need to enclose above status codes in double backticks as well.

Comment (CuriousLearner) [reply phone_verify/backends/base.py]: This only helps within the rST format, single backquotes are okay for comments

Comment (CuriousLearner) [reply phone_verify/backends/base.py]: Check out https://www.python.org/dev/peps/pep-0287/#id61

It states:

`single backquotes is recognized as "interpreted text", whose interpretation is application-dependent`

`double-backquotes to indicate program I/O or code snippets.`

**Pull request merged**